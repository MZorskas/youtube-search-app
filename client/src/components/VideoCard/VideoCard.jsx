import React, { memo } from 'react';
import './VideoCard.scss';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { LOGGER_ACTIONS } from 'loggerActions';

const VideoCard = memo(
  ({
    thumbnails,
    title,
    author,
    authorId,
    videoId,
    selectVideo,
    description,
    elementRef,
    logUserAction,
    className,
  }) => {
    const authorSrc = `https://www.youtube.com/channel/${authorId}`;

    const handleClick = () => {
      const videoDetails = {
        videoId: videoId,
        title: title,
        description: description,
        author: author,
        authorId: authorId,
      };

      selectVideo(videoDetails);
      logUserAction(LOGGER_ACTIONS.PREVIEW, videoDetails);
    };

    return (
      <div
        className={classNames('video-card', className)}
        data-testid="video-card"
        ref={elementRef}
      >
        <div
          className="video-card__image-wrapper"
          data-testid="video-card-image"
          onClick={handleClick}
        >
          <picture>
            <source
              media="(max-width: 500px)"
              srcSet={thumbnails.medium?.url}
              data-testid="video-card-source-1"
            />
            <source
              media="(min-width: 501px) and (max-width: 1200px)"
              srcSet={thumbnails.medium?.url}
              data-testid="video-card-source-2"
            />
            <source
              media="(min-width: 1201px)"
              srcSet={thumbnails.high?.url}
              data-testid="video-card-source-3"
            />
            <img
              src={thumbnails.high?.url}
              alt={title}
              className="video-card__image"
            />
          </picture>
        </div>
        <div className="video-card__info-wrapper">
          <h3 className="video-card__title" onClick={handleClick} tabIndex="0">
            {title}
          </h3>
          <a href={authorSrc} className="video-card__author" title={author}>
            {author}
          </a>
        </div>
      </div>
    );
  }
);

VideoCard.propTypes = {
  thumbnails: PropTypes.object.isRequired,
  title: PropTypes.string.isRequired,
  author: PropTypes.string.isRequired,
  authorId: PropTypes.string.isRequired,
  videoId: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  selectVideo: PropTypes.func.isRequired,
  elementRef: PropTypes.func,
  logUserAction: PropTypes.func.isRequired,
  className: PropTypes.string,
};

export default VideoCard;
