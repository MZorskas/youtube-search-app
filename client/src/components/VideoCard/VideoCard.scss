@import 'src/styles/mixins/index';
@import 'src/styles/base/base';

.video-card {
  display: flex;
  box-sizing: border-box;
  margin: 2rem 0;

  &__image-wrapper {
    flex: 1 0 50%;
    cursor: pointer;
  }

  &__info-wrapper {
    align-items: flex-start;
    box-sizing: border-box;
    display: flex;
    flex: 1 0 50%;
    flex-direction: column;
    padding-left: 1rem;
  }

  &__image {
    width: 100%;
    object-fit: cover;
  }

  &__title {
    @include text-style-h3;
    @include line-clamp(2);

    cursor: pointer;
    font-size: $font-size-large;
    padding: 0;
    margin: 0;
    word-break: break-word;
  }

  &__author {
    @include text-style-body;
    @include line-clamp(1);

    color: $color-font-secondary;
    padding-top: 1rem;
    margin: 0;
    text-decoration: none;
  }

  @include if-phone {
    &__title {
      @include text-style-h4;
    }
  }
}
