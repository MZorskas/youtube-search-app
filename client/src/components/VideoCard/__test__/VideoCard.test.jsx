import React from 'react';
import VideoCard from '../VideoCard';
import { render, fireEvent, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

const mockLogUserData = jest.fn();
const mockSelectVideo = jest.fn();

const thumbnails = {
  medium: {
    url: 'imgURL1',
  },
  high: {
    url: 'imgURL2',
  },
};

const videoDetails = {
  author: 'author',
  authorId: 'authorId',
  title: 'title',
  description: 'description',
  videoId: 'videoId',
};

const authorSrc = `https://www.youtube.com/channel/${videoDetails.authorId}`;

const videoCardSetup = () => {
  const utils = render(
    <VideoCard
      logUserAction={mockLogUserData}
      selectVideo={mockSelectVideo}
      thumbnails={thumbnails}
      author={videoDetails.author}
      authorId={videoDetails.authorId}
      title={videoDetails.title}
      description={videoDetails.description}
      videoId={videoDetails.videoId}
    />
  );
  const author = screen.getByText(videoDetails.author);
  const title = screen.getByText(videoDetails.title);
  const videoCard = screen.getByTestId('video-card');
  const image = screen.getByTestId('video-card-image');
  return {
    author,
    title,
    videoCard,
    image,
    ...utils,
  };
};

describe('VideoCard component', () => {
  it('renders with propTypes warning', () => {
    const spy = jest.spyOn(global.console, 'error');

    expect(screen.queryByTestId('video-card')).not.toBeInTheDocument();
    render(<VideoCard thumbnails={thumbnails} />);
    expect(screen.queryByTestId('video-card')).toBeInTheDocument();

    expect(spy).toHaveBeenCalled();
  });

  it('renders without propTypes warning', () => {
    const spy = jest.spyOn(global.console, 'error');

    expect(screen.queryByTestId('video-card')).not.toBeInTheDocument();
    render(
      <VideoCard
        logUserAction={mockLogUserData}
        selectVideo={mockSelectVideo}
        thumbnails={thumbnails}
        author={videoDetails.author}
        authorId={videoDetails.authorId}
        title={videoDetails.title}
        description={videoDetails.description}
        videoId={videoDetails.videoId}
      />
    );
    expect(screen.queryByTestId('video-card')).toBeInTheDocument();

    expect(spy).not.toHaveBeenCalled();
  });

  it('renders with correct prop values', () => {
    const { author, title, videoCard } = videoCardSetup();
    expect(videoCard).toBeInTheDocument();
    expect(author).toBeTruthy();
    expect(title).toBeTruthy();
    expect(
      screen.getByTestId('video-card-source-1').getAttribute('srcSet')
    ).toMatch(thumbnails.medium.url);
    expect(
      screen.getByTestId('video-card-source-2').getAttribute('srcSet')
    ).toMatch(thumbnails.medium.url);

    expect(
      screen.getByTestId('video-card-source-3').getAttribute('srcSet')
    ).toMatch(thumbnails.high.url);
    expect(screen.getByAltText(videoDetails.title).getAttribute('src')).toMatch(
      thumbnails.high.url
    );
    expect(screen.getByTitle(videoDetails.author).getAttribute('href')).toBe(
      authorSrc
    );
  });

  it('selects video on title click', () => {
    const { title } = videoCardSetup();
    expect(mockSelectVideo).toHaveBeenCalledTimes(0);

    fireEvent.click(title);
    expect(mockSelectVideo).toHaveBeenCalledWith(videoDetails);
    expect(mockSelectVideo).toHaveBeenCalledTimes(1);
  });

  it('selects video on image click', () => {
    const { image } = videoCardSetup();
    expect(mockSelectVideo).toHaveBeenCalledTimes(0);

    fireEvent.click(image);
    expect(mockSelectVideo).toHaveBeenCalledWith(videoDetails);
    expect(mockSelectVideo).toHaveBeenCalledTimes(1);
  });

  it('logs user action on title click', () => {
    const { title } = videoCardSetup();
    expect(mockLogUserData).toHaveBeenCalledTimes(0);

    fireEvent.click(title);
    expect(mockLogUserData).toHaveBeenCalledTimes(1);
  });

  it('logs user action on image click', () => {
    const { image } = videoCardSetup();
    expect(mockLogUserData).toHaveBeenCalledTimes(0);

    fireEvent.click(image);
    expect(mockLogUserData).toHaveBeenCalledTimes(1);
  });

  //   it('renders with additional className', () => {
  //     expect(screen.queryByTestId('input')).not.toBeInTheDocument();
  //     render(
  //       <Input
  //         onChange={mockOnChange}
  //         className="input-test"
  //         placeholder="Gifs"
  //       />
  //     );
  //     expect(screen.getByPlaceholderText('Gifs').className).toBe(
  //       'input-wrapper__input input-test'
  //     );
  //   });
  //   it('updates input value', () => {
  //     const { input } = inputSetup();
  //     fireEvent.change(input, { target: { value: '123' } });
  //     expect(input.value).toBe('123');
  //   });
  //   it('removes input value', () => {
  //     const { input } = inputSetup();
  //     fireEvent.change(input, { target: { value: '123' } });
  //     expect(input.value).toBe('123');
  //     fireEvent.change(input, { target: { value: '' } });
  //     expect(input.value).toBe('');
  //   });
  //   it('matches snapshot', () => {
  //     const tree = renderer
  //       .create(
  //         <Input
  //           onChange={mockOnChange}
  //           className="input-test"
  //           placeholder="Gifs"
  //           id="search-input"
  //           label="Search"
  //           autoFocus
  //         />
  //       )
  //       .toJSON();
  //     expect(tree).toMatchSnapshot();
  //   });
  it('matches snapshot', () => {
    const tree = renderer
      .create(
        <VideoCard
          logUserAction={mockLogUserData}
          selectVideo={mockSelectVideo}
          thumbnails={thumbnails}
          author={videoDetails.author}
          authorId={videoDetails.authorId}
          title={videoDetails.title}
          description={videoDetails.description}
          videoId={videoDetails.videoId}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
