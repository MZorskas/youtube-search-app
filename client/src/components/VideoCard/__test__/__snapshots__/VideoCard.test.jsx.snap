// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoCard component matches snapshot 1`] = `
<div
  className="video-card"
  data-testid="video-card"
>
  <div
    className="video-card__image-wrapper"
    data-testid="video-card-image"
    onClick={[Function]}
  >
    <picture>
      <source
        data-testid="video-card-source-1"
        media="(max-width: 500px)"
        srcSet="imgURL1"
      />
      <source
        data-testid="video-card-source-2"
        media="(min-width: 501px) and (max-width: 1200px)"
        srcSet="imgURL1"
      />
      <source
        data-testid="video-card-source-3"
        media="(min-width: 1201px)"
        srcSet="imgURL2"
      />
      <img
        alt="title"
        className="video-card__image"
        src="imgURL2"
      />
    </picture>
  </div>
  <div
    className="video-card__info-wrapper"
  >
    <h3
      className="video-card__title"
      onClick={[Function]}
      tabIndex="0"
    >
      title
    </h3>
    <a
      className="video-card__author"
      href="https://www.youtube.com/channel/authorId"
      title="author"
    >
      author
    </a>
  </div>
</div>
`;
